core/src/main/java/fr/sncf/osrd/utils/new_graph/Pathfinding.java

```diff
@@ -0,0 +1,179 @@
+    ) {}
+
+    /** A range, made of edge + start and end offsets on the edge. Used for the output of the pathfinding */
+    public record EdgeRange<EdgeT>(
+            EdgeT edge,
+            double start,
+            double end
+    ) {}
+
+    /** Step priority queue */
+    private final PriorityQueue<Step<EdgeT>> queue;
+    /** Function to call to get edge length */
+    private final Function<EdgeT, Double> edgeToLength;
+    /** Input graph */
+    private final ImmutableNetwork<NodeT, EdgeT> graph;
+    /** Keeps track of visited location. For each visited range, keeps the max number of passed steps at that point */
+    private final HashMap<EdgeRange<EdgeT>, Integer> seen = new HashMap<>();
+    // TODO: add a `private final Function<EdgeT, Boolean> isEdgeAllowed;` as optional parameter to filter some edges
+
+    /** Returns the shortest path from any start edge to any end edge, as a list of edge (containing start and stop) */
+    public static <NodeT, EdgeT> List<EdgeT> findEdgePath(
+            ImmutableNetwork<NodeT, EdgeT> graph,
+            List<Collection<EdgeLocation<EdgeT>>> locations,
+            Function<EdgeT, Double> edgeToLength
+    ) {
+        return new Pathfinding<>(graph, edgeToLength).runPathfindingEdgesOnly(locations);
+    }
+
+    /** Returns the shortest path from any start edge to any end edge, as a list of (edge, start, end) */
+    public static <NodeT, EdgeT> List<EdgeRange<EdgeT>> findPath(
+            ImmutableNetwork<NodeT, EdgeT> graph,
+            List<Collection<EdgeLocation<EdgeT>>> locations,
+            Function<EdgeT, Double> edgeToLength
+    ) {
+        return new Pathfinding<>(graph, edgeToLength).runPathfinding(locations);
+    }
+
+    /** Constructor */
+    private Pathfinding(
+            ImmutableNetwork<NodeT, EdgeT> graph,
+            Function<EdgeT, Double> edgeToLength
+    ) {
+        queue = new PriorityQueue<>();
+        this.graph = graph;
+        this.edgeToLength = edgeToLength;
+    }
+
+    /** Runs the pathfinding, returning a path as a list of (edge, start offset, end offset).
+     * Each step is given as a collection of location.
+     * It finds the shortest path from start to end, going through every intermediate step in order.
+     * It uses Dijkstra algorithm internally, but can be changed to an A* with minor changes */
+    private List<EdgeRange<EdgeT>> runPathfinding(
+            List<Collection<EdgeLocation<EdgeT>>> locations
+    ) {
+        for (var location : locations.get(0)) {
+            var startRange = new EdgeRange<>(location.edge, location.offset, edgeToLength.apply(location.edge));
+            registerStep(startRange, null, 0, 0);
+        }
+        while (true) {
+            var step = queue.poll();
+            if (step == null)
+                return null;
+            if (hasReachedEnd(locations, step))
+                return buildResult(step);
+            for (var stopLocation : locations.get(step.nReachedSteps + 1))
```

Comment:
it'd be nice to have a little comment above this line, like: "check if this step reaches the current target"